src/frontend/src/handler/alter_table_with_sr.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use pgwire::pg_response::StatementType;
+use risingwave_common::bail_not_implemented;
+use risingwave_common::types::DataType;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+use risingwave_sqlparser::ast::{ConnectorSchema, ObjectName, Statement};
+use risingwave_sqlparser::parser::Parser;
+
+use super::alter_source_with_sr::alter_definition_format_encode;
+use super::alter_table_column::{fetch_table_catalog_for_alter, schema_has_schema_registry};
+use super::{HandlerArgs, RwPgResponse};
+use crate::error::{ErrorCode, Result, RwError};
+use crate::handler::alter_source_with_sr::{
+    check_format_encode, fetch_source_catalog_with_db_schema_id,
+};
+use crate::handler::create_table::{generate_stream_graph_for_table, ColumnIdGenerator};
+
+pub async fn handle_alter_table_with_sr(
+    handler_args: HandlerArgs,
+    name: ObjectName,
+    connector_schema: ConnectorSchema,
+) -> Result<RwPgResponse> {
+    let session = handler_args.session;
+    let original_table = fetch_table_catalog_for_alter(session.as_ref(), &name)?;
+    let (original_source, _, _) = fetch_source_catalog_with_db_schema_id(session.as_ref(), &name)?;
+
+    if !original_table.incoming_sinks.is_empty() {
+        bail_not_implemented!("alter table with incoming sinks");
+    }
+
+    // TODO(yuhao): alter table with generated columns.
+    if original_table.has_generated_column() {
+        return Err(RwError::from(ErrorCode::BindError(
+            "Alter a table with generated column has not been implemented.".to_string(),
+        )));
+    }
+
+    check_format_encode(&original_source, &connector_schema)?;
+
+    if !schema_has_schema_registry(&connector_schema) {
+        return Err(ErrorCode::NotSupported(
+            "altering a table without schema registry".to_string(),
+            "try `ALTER TABLE .. ADD COLUMN ...` instead".to_string(),
+        )
+        .into());
+    }
+
+    let definition = alter_definition_format_encode(
+        &original_table.definition,
+        connector_schema.row_options.clone(),
+    )?;
+
+    let [definition]: [_; 1] = Parser::parse_sql(&definition)
+        .context("unable to parse original table definition")?
+        .try_into()
+        .unwrap();
```

Comment:
Why `unwrap()` instead of `return Err`?

Reply:
Because the definition is converted from SQL AST in `alter_definition_format_encode`, and should be error-free as converted back.

Reply:
I think the code here is directly duplicated from `alter_table_column`. Could you please try your best to reuse some code snippets between them? Otherwise, it can be painful to maintain.